albatross.ml

```diff
@@ -533,82 +552,127 @@ module Make (S : Tcpip.Stack.V4V6) = struct
                                "albatross received error reading from %a \
                                 querying %a %a: %a"
                                Fmt.(pair ~sep:(any ":") Ipaddr.pp int)
-                               t.remote Vmm_core.Name.pp name
+                               (config.server_ip, config.server_port)
+                               Vmm_core.Name.pp name
                                (Vmm_commands.pp ~verbose:false)
                                cmd TLS.pp_error e))
                   | Error err -> Lwt.return_error err)))
 
-  let init stack server ?(port = 1025) cert key =
+  let init stack (configuration : Configuration.t) =
     let open Lwt.Infix in
-    let t =
-      { stack; remote = (server, port); cert; key; policies = Vmm_trie.empty }
-    in
     let cmd = `Policy_cmd `Policy_info in
-    match gen_cert t cmd "." with
-    | Error s -> invalid_arg s
-    | Ok certificates -> (
-        raw_query t certificates cmd reply >|= function
+    let name_for_cert =
+      Vmm_core.Name.of_string configuration.name |> Result.get_ok
+    in
+    match
+      key_cert ~is_ca:false ~cmd configuration.private_key configuration.name
+        (X509.Certificate.subject configuration.certificate)
+    with
+    | Error s ->
+        Logs.warn (fun m ->
+            m "init: key_cert failed for %s: %s" configuration.name s);
+        let state =
+          {
+            errors = Some "Initialization error with key certificate";
+            policies = Vmm_trie.empty;
+            configuration;
+          }
+        in
+        Lwt.return state
+    | Ok (key, cert) -> (
+        let certificates = `Single ([ cert ], key) in
+        raw_query stack configuration ~name:name_for_cert certificates cmd reply
+        >>= function
         | Ok (_hdr, `Success (`Policies ps)) ->
-            let ps =
+            let policies =
               List.fold_left
-                (fun acc (name, p) -> fst (Vmm_trie.insert name p acc))
+                (fun trie (name, p) -> fst (Vmm_trie.insert name p trie))
                 Vmm_trie.empty ps
             in
-            t.policies <- ps;
-            t
-        | Ok w ->
-            Logs.err (fun m ->
-                m "albatross expected success policies, got reply %a"
-                  (Vmm_commands.pp_wire ~verbose:false)
-                  w);
-            t
+            let state = { errors = None; policies; configuration } in
+            Lwt.return state
+        | Ok _w ->
+            Logs.warn (fun m ->
+                m "init: unexpected policy reply for %s" configuration.name);
+            let state =
+              {
+                errors = Some "Unexpected policy reply";
```

Comment:
This is a remote error (unexpected reply), again maybe worth to show what we received?